### PR TITLE
Added SaveBackup/RestoreBackup

### PR DESCRIPTION
**Pull request by cielavenir**
(Sorry for WIP PR because I need to test on actual controller. dev-machine test is already done)
(As we don't have so much time, I need to confirm about API interface urgently)

### CODE REVIEW

**Comment by cielavenir** [comment]
Tested on actual controller.

**Comment by ziyan** [comment]
1. please add comments about new apis
2. please increment minor version of library
3. is there any library to construct query string in a not so manual way?

**Comment by cielavenir** [comment]
1 and 2 are done.

**Comment by cielavenir** [comment]
for 3, do we have our standard?

**Comment by cielavenir** [comment]
@ziyan

**Comment by cielavenir** [comment]
Ziyan, we are in hurry. Now I need to invite Kei...

File: samples/mujinbackup.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    const unsigned int useconfig = opts["useconfig"].as<unsigned int>();
+    const unsigned int usemedia = opts["usemedia"].as<unsigned int>();
+    stringstream urlss;
+    urlss << "http://" << hostname << ":" << controllerPort;
+
+    // connect to mujin controller
+    ControllerClientPtr controllerclient = CreateControllerClient(controllerUsernamePass, urlss.str());
+    cerr << "connected to mujin controller at " << urlss.str() << endl;
+
+    if(!isatty(fileno(stdout))){
```
Comment (ziyan) [inline samples/mujinbackup.cpp]: This is quite magical, will the customer understand this usage?

File: src/controllerclientimpl.h [hunk 2]
```diff
@@ -216,6 +220,7 @@ class ControllerClientImpl : public ControllerClient, public boost::enable_share
     ///
     /// overwrites the file if it already exists.
     virtual void _UploadFileToControllerViaForm(const std::string& file, const std::string& filename);
+    virtual void _UploadFileToControllerViaForm2(const std::string& file, const std::string& filename, const std::string& endpoint);
```
Comment (ziyan) [inline src/controllerclientimpl.h]: There are only 2 places that use `_UploadFileToControllerViaForm`, please change that function and have the caller call with endpoint instead.

File: src/controllerclientimpl.h [hunk 3]
```diff
@@ -64,6 +64,10 @@ class ControllerClientImpl : public ControllerClient, public boost::enable_share
     virtual void DeleteDirectoryOnController_UTF8(const std::string& desturi);
     virtual void DeleteDirectoryOnController_UTF16(const std::wstring& desturi);
 
+    virtual void SaveBackup(std::vector<unsigned char>& vdata, bool config, bool media, double timeout);
+    virtual void RestoreBackup_UTF8(const std::string& filename_utf8, bool config, bool media);
```
Comment (ziyan) [inline src/controllerclientimpl.h]: I think going forward, we should avoid forcing user to pass in filename, instead just pass in a stream or vector of bytes. You also don't need two functions in that case. And you have parity between SaveBackup and RestoreBackup.

**Comment by cielavenir** [comment]
@ziyan Updated. Please check again.

File: src/controllerclientimpl.h [hunk 4]
```diff
@@ -64,6 +64,9 @@ class ControllerClientImpl : public ControllerClient, public boost::enable_share
     virtual void DeleteDirectoryOnController_UTF8(const std::string& desturi);
     virtual void DeleteDirectoryOnController_UTF16(const std::wstring& desturi);
 
+    virtual void SaveBackup(std::vector<unsigned char>& vdata, bool config, bool media, double timeout);
+    virtual void RestoreBackup(const std::vector<unsigned char>& vdata, bool config, bool media);
```
Comment (ziyan) [inline src/controllerclientimpl.h]:
```
virtual void SaveBackup(std::ostream& outputStream, bool config, bool media, double timeout);
virtual void RestoreBackup(const std::istream& inputStream, bool config, bool media);
```

User could open file as stream, or use in memory buffer

```
std::ofstream ofs("backup.tar", std::ofstream::out);
```

Comment (cielavenir) [reply src/controllerclientimpl.h]: I have written the following script:

```
#!/usr/bin/python
import socket
sock = socket.socket()
sock.connect(('127.0.0.1',80))
if 1:
    sock.send('''POST /fileupload HTTP/1.1
Host: localhost
X-CSRFToken: csrftoken
Accept: */*
User-Agent: upload_test
Connection: close
Referer: http://localhost
Cookie: csrftoken=csrftoken
Content-Type: multipart/form-data; boundary=section
Authorization: Basic bXVqaW46bXVqaW4=
Transfer-Encoding: chunked

5b
--section
Content-Disposition: form-data; name="file"; filename="file"

abcdefghijklmn
1b
opqrstuvwxyz
--section--
0
'''.replace('\n','\r\n'))
else:
    sock.send('''POST /fileupload HTTP/1.1
Host: localhost
X-CSRFToken: csrftoken
Accept: */*
User-Agent: upload_test
Connection: close
Referer: http://localhost
Cookie: csrftoken=csrftoken
Content-Type: multipart/form-data; boundary=section
Authorization: Basic bXVqaW46bXVqaW4=
Content-Length: 118

--section
Content-Disposition: form-data; name="file"; filename="file"

abcdefghijklmn
opqrstuvwxyz
--section--
'''.replace('\n','\r\n'))
print sock.recv(4096)
```

When I disable chunk, I can upload the file. However when I enable chuck, firstly webstack returned the following error:

```
411 Length Required
A request of the requested method POST requires a valid Content-length.
```

stackoverflow says I can put `WSGIChunkedRequest On`, so I retried. Then I got internal server error with the error:

```
Request data read error when proxying data to daemon process: Unknown error -102.
```

Now I need to say that I need to precompute Content-Length (i.e. CURLFORM_CONTENTSLENGTH). So istream cannot be used and ifstream needs to be used.
For controllerclientpy, https://github.com/psf/requests/blob/master/requests/models.py#L159 indicates that they are reading whole file to compute Content-Length.
For ostream I don't have such concerns yet.

Comment (ziyan) [reply src/controllerclientimpl.h]: What's the unknown error -102? Any log message  in webstack.log?

Comment (ziyan) [reply src/controllerclientimpl.h]: Also, istream doesn't have seekg and tellg?

Comment (cielavenir) [reply src/controllerclientimpl.h]: > unknown error -102

that was all webstack.log printed.

> istream doesn't have seekg and tellg

sure, istream does support them. the caveat is (in `controllerclient->RestoreBackup(cin)`) when cin is '< input.tar', seek will success, but when cin is `cat input.tar |`, seek will fail. I added error handling and document.

So, restoring concern is now cleared, right?

Comment (ziyan) [reply src/controllerclientimpl.h]: Right, but you didn't fix `SaveBackup`, `std::ostream& outputStream` they should be symmetric

File: src/controllerclientimpl.cpp [hunk 5]
```diff
@@ -1575,30 +1619,40 @@ void ControllerClientImpl::_UploadFileToController(FILE* fd, const std::string&
...
     CURL_FORM_RELEASER(formpost);
     curl_formadd(&formpost, &lastptr,
                  CURLFORM_COPYNAME, "files[]",
-                 CURLFORM_FILE, file.c_str(),
+                 CURLFORM_FILENAME, filename.c_str(),
+                 CURLFORM_STREAM, &inputStream,
+                 CURLFORM_CONTENTSLENGTH, contentLength,
                  CURLFORM_END);
     curl_formadd(&formpost, &lastptr,
                  CURLFORM_COPYNAME, "filename",
```
Comment (ziyan) [inline src/controllerclientimpl.cpp]: Hmm, why is this here? "filename" for backup restore endpoint does not make sense.

File: src/controllerclientimpl.cpp [hunk 6]
```diff
@@ -1575,30 +1619,40 @@ void ControllerClientImpl::_UploadFileToController(FILE* fd, const std::string&
...
+    }
+    std::cout << contentLength << std::endl;
+
+    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_READFUNCTION, NULL, _ReadIStreamCallback);
     // prepare form
     struct curl_httppost *formpost = NULL;
     struct curl_httppost *lastptr = NULL;
     CURL_FORM_RELEASER(formpost);
     curl_formadd(&formpost, &lastptr,
                  CURLFORM_COPYNAME, "files[]",
```
Comment (ziyan) [inline src/controllerclientimpl.cpp]: Maybe this should be a param of the function.

Comment (cielavenir) [reply src/controllerclientimpl.cpp]: Could you explain what do you mean by "this"?

File: src/controllerclientimpl.cpp (hunk 5 above)
Comment (cielavenir) [reply src/controllerclientimpl.cpp]: because... _UploadFileToControllerViaForm is shared among UploadFileToController and RestoreBackup... I added some hacks in 3b407da ...

**Comment by cielavenir** [comment]
@ziyan Talked with Kei and he said leaving HTTP error json only to outputStream is not best idea.

I tried to address this as https://github.com/mujin/controllerclientcpp/commit/2da9ea727c597316c589c89b61afb0152ba60b69 , but this is too cryptic and put to another branch...

**Comment by cielavenir** [comment]
Also I tested restoring tar.bz2 (as well as tar), though I was told not to document it.

**Comment by rdiankov** [comment]
looks good enough for now, thanks!
